prow/tide/tide.go

```diff
@@ -573,6 +581,90 @@ func filterPR(ghc githubClient, sp *subpool, pr *PullRequest) bool {
 	return false
 }
 
+// mergeMethodChecker provides a function to check if a PR can be merged with
+// the requested method and does not have a merge conflict.
+// It caches results and should be cleared periodically with clearCache()
+type mergeMethodChecker struct {
+	config config.Getter
+	ghc    githubClient
+	log    *logrus.Entry
+
+	sync.Mutex
+	cache map[config.OrgRepo]map[github.PullRequestMergeType]bool
+}
+
+func newMergeMethodChecker(cfg config.Getter, ghc githubClient, log *logrus.Entry) *mergeMethodChecker {
+	m := &mergeMethodChecker{
+		config: cfg,
+		ghc:    ghc,
+		log:    log,
+		cache:  map[config.OrgRepo]map[github.PullRequestMergeType]bool{},
+	}
+
+	go m.clearCache()
+	return m
+}
+
+func (m *mergeMethodChecker) clearCache() {
+	// Only do this once per token reset since it could be a bit expensive for
+	// Tide instances that handle hundreds of repos.
+	ticker := time.NewTicker(time.Hour)
+	for {
+		<-ticker.C
+		m.Lock()
+		m.cache = make(map[config.OrgRepo]map[github.PullRequestMergeType]bool)
```

Comment:
ghproxy already knows how to cache and can also check if its cache is still valid. Why do we need a second caching layer here?

Reply:
If we don't do any in-process caching this would require 1-2 extra API requests every sync period for every PR that matches a Tide query and an additional request whenever an open PR outside the tide pool is updated. Thats a pretty significant token increase that would make ghproxy required for anything, but small Tide instances if other Prow components are running.
We've been trying to keep ghproxy as an optional component that only large Prow instances need to use, but perhaps it is time to reconsider that. It seems relatively reliable and would make token management easier. We should consider making it HA before we do that though.

Reply:
So the reason I asked about that is that I dislike both the added complexity in the code itself and when trying to debug this. Also, I think this is pretty subpar for users as well, because if someone does change their allowed mergeMethods, they have to wait for up to one hour without getting any feedback on why Tide tells them their newly-configured mergeMethod wasn't allowed.

Maybe a middleground would be to just cache during the sync?

> We've been trying to keep ghproxy as an optional component that only large Prow instances need to use, but perhaps it is time to reconsider that. It seems relatively reliable and would make token management easier. 

Yeah, IMHO this would simplify things a lot because we do not have to think anymore about how many requests we do, but just about how many requests we do that may not be served from cache. Also, I don't think we even have data on the token usage of a small prow without ghproxy, so we can not even say if we support that usecase well. On top of that even when keeping the token issue aside, ghproxy also speeds things up.

> We should consider making it HA before we do that though.

Definitely +1 for making it HA (we had a brief outage yesterday due to it going down) but I don't think that should be a requirement for making it mandatory.

Reply:
> Also, I think this is pretty subpar for users as well, because if someone does change their allowed mergeMethods, they have to wait for up to one hour without getting any feedback on why Tide tells them their newly-configured mergeMethod wasn't allowed.

Thats fair, but as noted in the PR description this rarely, if ever, changes so I figured a delay was worth the benefits (especially if we made the delay configurable). It is also unclear to me why this would be difficult to debug. I think the logs and errors are informative and debugging a single process would be easier than debugging the interaction with ghproxy.

> I don't think we even have data on the token usage of a small prow without ghproxy, so we can not even say if we support that usecase well.

Well I don't have hard data from a small prow, but given that k8s Prow (very large) worked without ghproxy and that we have been careful to avoid relying on its existence, I don't know what would have changed to make it impossible to run Tide without ghproxy at reasonable scale.

I agree that assuming ghproxy is in use simplifies the code. But it does so in exchange for extra network traffic (not a big issue, but more errors), a hard requirement to use ghproxy (maybe not an issue, but a breaking change), and a single point of failure (we fall back to using the GH API directly, but that would cause us to quickly exhaust tokens without caching in process). The last point is the main one that concerns me, especially if it caused an outage for you as recently as yesterday.

I'm happy to change this if we want to start requiring ghproxy, but I'd like to get this functionality in and not block it on that policy change (since it is a breaking change we need a warning period where we warn if it looks like components are using the GH API directly). There are a number of places where we do similar in-process caching because we don't want to assume ghproxy exists so this isn't a problem or pattern unique to this PR.

Reply:
> Thats fair, but as noted in the PR description this rarely, if ever, changes so I figured a delay was worth the benefits (especially if we made the delay configurable). It is also unclear to me why this would be difficult to debug. I think the logs and errors are informative and debugging a single process would be easier than debugging the interaction with ghproxy.

Well, because if an allowed merge method is refused due to cache staleness, ppl essentially have to look at the code and find this particular cache in order to understand why that happens.

> There are a number of places where we do similar in-process caching because we don't want to assume ghproxy exists so this isn't a problem or pattern unique to this PR.

Do you have an example at hand for this? I am not aware of other places where there is up to an hour waiting time for a cache to not be stale anymore.

Reply:
I don't have an example where we only update the cache once a sync period, but that is because every other cache is for relatively dynamic data, not something as static as the allowed merge methods which don't ever change for most repos. It is certainly possible that someone could change the allowed merge methods and be confused why there is a delay before Tide reacts, but I don't think that will be a significant problem in practice. Currently we see entire repos get blocked by this and typically noone notices until we see the errors in the Tide logs.

More importantly, even if there is a one time delay if settings are changed this is still a big improvement because it unblocks the repos and I'd rather not wait months for a warning period to enable ghproxy when we could merge this today.
I agree that this isn't perfect and requiring ghproxy is the end goal, but wouldn't we rather have this in the meantime? In particular I'm trying to make this PR safe: https://github.com/kubernetes/test-infra/pull/16160

Reply:
Do we know anyone running prow without ghproxy?

Reply:
Yes. Looks like OSS Prow doesn't have it running and Istio didn't have ghproxy until last July. Looking around at some other instances we know of I see that Knative, Azure aks-engine, and GoogleCloudPlatform/compute-image-tools instances also do not have it deployed.

Reply:
Okay, so i continue to not really like this since understanding why a PR gets refused while the cache is stale basically requires to be aware of this cache which none of our users will be, but I do realize that having tide blocked by an unmergeable PR is the bigger evil. Can we create an issue to have this cache invalidated after each sync loop after may first (xref https://github.com/kubernetes/test-infra/pull/16763)?

/lgtm
/approve
/hold

Reply:
Opened https://github.com/kubernetes/test-infra/issues/16800.
I definitely agree that this is the lesser of two evils, but still an evil and I am eager to exorcise when we have all Prow's using ghProxy.
Thanks for the review Alvaro!